# coding: utf8
############################################################################
#
# rtsp_image_transport
# Copyright © 2021 Fraunhofer FKIE
# Author: Timo Röhling
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
############################################################################
#
from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# The level encodes which parts of the subscriber need to be reinitialized
LVL_CODEC = 1
LVL_SESSION = 2
LVL_SERVER = 4

reconnect_enum = gen.enum(
    [
        gen.const("ReconnectNever", int_t, 0, "Never"),
        gen.const("ReconnectOnTimeout", int_t, 1, "On Timeout"),
        gen.const("ReconnectOnFailure", int_t, 2, "On Failure"),
        gen.const("ReconnectAlways", int_t, 3, "Always"),
    ], "Reconnect policies")

gen.add("use_hw_decoder", bool_t, LVL_CODEC, "Use NVDEC hardware acceleration if possible", False)
gen.add("reconnect_policy", int_t, 0, "Reconnect to RTSP server", 1, 0, 3, edit_method=reconnect_enum)
gen.add("timeout", double_t, 0, "Session timeout (seconds)", 2.0, 0.0, 60.0)
gen.add("reconnect_minwait", double_t, 0, "Minimum delay between connection atttempts (seconds)", 0.1, 0.0, 60.0)
gen.add("reconnect_maxwait", double_t, 0, "Maximum delay between connection attempts (seconds)", 30.0, 0.0, 600.0)

gen.generate("rtsp_image_transport", "RTSPSubscriber", "RTSPSubscriber")
